drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1034 @@
+	data->ecc_layout_pos = 2; /* skip the bad block mark for Macronix spi nand */
+
+	data->nbc.bch = bch_init(m, t);
+	if (!data->nbc.bch) {
+		return;
+	}
+
+	/* verify that eccbytes has the expected value */
+	if (data->nbc.bch->ecc_words * 4 != eccbytes) {
+		LOG_ERR("invalid eccbytes %u, should be %u\n",
+			eccbytes, data->nbc.bch->ecc_words);
+		return;
+	}
+
+	data->page_buf = (uint8_t *)k_malloc(data->page_size + data->oob_size);
+	if (data->page_buf == NULL) {
+		LOG_ERR("Not enougn heap \n");
+	}
+	data->ecc_calc = (uint8_t *)k_malloc(data->ecc_steps * data->ecc_bytes);
+	if (data->ecc_calc == NULL) {
+		LOG_ERR("Not enougn heap \n");
+	}
+	data->ecc_code = (uint8_t *)k_malloc(data->ecc_steps * data->ecc_bytes);
+	if (data->ecc_code == NULL) {
+		LOG_ERR("Not enougn heap \n");
+	}
+	data->nbc.eccmask = (unsigned char *)k_malloc(eccbytes);
+
+	data->nbc.errloc = (unsigned int *)k_malloc(t * sizeof(*data->nbc.errloc));
+	if (!data->nbc.eccmask || !data->nbc.errloc) {
+		LOG_ERR("Not enougn heap \n");
+		return;
+	}
+	/*
+	* compute and store the inverted ecc of an erased ecc block
+	*/
+	erased_page = (unsigned char *)k_malloc(eccsize);
+	if (!erased_page) {
+		return;
+	}
+	memset(data->page_buf, 0xff, data->page_size + data->oob_size);
+	memset(erased_page, 0xff, eccsize);
+	memset(data->nbc.eccmask, 0, eccbytes);
+	bch_encode(data->nbc.bch, erased_page, (unsigned int *)data->nbc.eccmask);
+	k_free(erased_page);
+
+	for (i = 0; i < eccbytes; i++) {
+		data->nbc.eccmask[i] ^= 0xff;
+	}
+}
+
+void bch_ecc_free(const struct device *dev)
+{
+	struct spi_nand_data *data = dev->data;
+	bch_free(data->nbc.bch);
+	k_free(data->nbc.errloc);
+	k_free(data->nbc.eccmask);
+	k_free(data->page_buf);
+	k_free(data->ecc_calc);
+	k_free(data->ecc_code);
+}
+
+static int spi_nand_read_otp_onfi(const struct device *dev)
+{
+	int ret;
+	uint8_t secur_reg = 0, onfi_table[256];
+	struct spi_nand_data *data = dev->data;
+
+	data->page_shift = 12;
+	data->page_size = 2048;
+
+	acquire_device(dev);
+
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out0;
+	}
+
+	secur_reg |= SPINAND_SECURE_BIT_OTP_EN;
+	ret = spi_nand_set_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, secur_reg);
+	if (ret != 0) {
+		LOG_ERR("set feature failed: %d", ret);
+		goto out0;
+	}
+
+	ret = spi_nand_get_feature(dev, SPI_NAND_FEA_ADDR_CONF_B0, &secur_reg);
+	if (ret != 0) {
+		LOG_ERR("get feature failed: %d", ret);
+		goto out0;
+	}
+
+	if (!(secur_reg & SPINAND_SECURE_BIT_OTP_EN)) {
+		LOG_ERR("Enable OTP failed: %d\n", secur_reg);
+	}
+
+out0:	release_device(dev);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = spi_nand_read(dev, 1 << data->page_shift, onfi_table, sizeof(onfi_table));
+	if (ret != 0) {
+		LOG_ERR("read onfi table failed: %d", ret);
+		return ret;
+	}
+
+	if (onfi_table[0] == 'O' && onfi_table[1] == 'N'  
+		&& onfi_table[2] == 'F' && onfi_table[3] == 'I') {
```

Comment:
The last 2 bytes of the 256-byte table is a CRC calculated across the table. This should be checked prior to reading other parameters from the table.

The full table contents can be found here: https://www.onfi.org/-/media/client/onfi/specs/onfi_2_0_gold.pdf?la=en
in section 5.6.1. The CRC is described in section 5.6.1.43 and sample code is provided in Appendix A.